File: frontend/lib/api/skills.ts
/**
 * 技能管理 API 客户端
 */

import { apiRequest } from "@/lib/api/client";

// ========== 类型定义 ==========

export type SkillType = "system" | "user" | "ai";
export type SkillCategory = "prompt" | "retrieval" | "tool" | "workflow";

export interface Skill {
  id: string;
  name: string;
  description: string;
  category: SkillCategory;
  content: string;
  trigger_keywords: string[];
  trigger_intents: string[];
  always_apply: boolean;
  applicable_agents: string[];
  type: SkillType;
  version: string;
  author: string | null;
  is_active: boolean;
  is_system: boolean;
  created_at: string;
  updated_at: string;
}

export interface SkillListResponse {
  items: Skill[];
  total: number;
  page: number;
  page_size: number;
}

export interface SkillCreate {
  name: string;
  description: string;
  category?: SkillCategory;
  content: string;
  trigger_keywords?: string[];
  trigger_intents?: string[];
  always_apply?: boolean;
  applicable_agents?: string[];
}

export interface SkillUpdate {
  name?: string;
  description?: string;
  category?: SkillCategory;
  content?: string;
  trigger_keywords?: string[];
  trigger_intents?: string[];
  always_apply?: boolean;
  applicable_agents?: string[];
  is_active?: boolean;
}

export interface SkillGenerateRequest {
  description: string;
  category?: SkillCategory;
  applicable_agents?: string[];
  examples?: string[];
}

export interface SkillGenerateResponse {
  skill: SkillCreate;
  confidence: number;
  suggestions: string[];
}

export interface AgentSkillConfig {
  skill_id: string;
  priority: number;
  is_enabled: boolean;
}

export interface AgentSkillRead {
  skill_id: string;
  skill_name: string;
  skill_description: string;
  priority: number;
  is_enabled: boolean;
}

// ========== API 函数 ==========

const BASE_URL = "/api/v1/admin/skills";

/**
 * 获取技能列表
 */
export async function listSkills(params?: {
  type?: SkillType;
  category?: SkillCategory;
  is_active?: boolean;
  page?: number;
  page_size?: number;
}): Promise<SkillListResponse> {
  const searchParams = new URLSearchParams();
  if (params?.type) searchParams.set("type", params.type);
  if (params?.category) searchParams.set("category", params.category);
  if (params?.is_active !== undefined)
    searchParams.set("is_active", String(params.is_active));
  if (params?.page) searchParams.set("page", String(params.page));
  if (params?.page_size) searchParams.set("page_size", String(params.page_size));

  const query = searchParams.toString();
  const url = query ? `${BASE_URL}?${query}` : BASE_URL;
  return apiRequest<SkillListResponse>(url);
}

/**
 * 获取技能详情
 */
export async function getSkill(skillId: string): Promise<Skill> {
  return apiRequest<Skill>(`${BASE_URL}/${skillId}`);
}

/**
 * 创建技能
 */
export async function createSkill(data: SkillCreate): Promise<Skill> {
  return apiRequest<Skill>(BASE_URL, {
    method: "POST",
    body: JSON.stringify(data),
  });
}

/**
 * 更新技能
 */
export async function updateSkill(
  skillId: string,
  data: SkillUpdate
): Promise<Skill> {
  return apiRequest<Skill>(`${BASE_URL}/${skillId}`, {
    method: "PUT",
    body: JSON.stringify(data),
  });
}

/**
 * 删除技能
 */
export async function deleteSkill(skillId: string): Promise<void> {
  await apiRequest<void>(`${BASE_URL}/${skillId}`, {
    method: "DELETE",
  });
}

/**
 * AI 生成技能
 */
export async function generateSkill(
  data: SkillGenerateRequest
): Promise<SkillGenerateResponse> {
  return apiRequest<SkillGenerateResponse>(`${BASE_URL}/generate`, {
    method: "POST",
    body: JSON.stringify(data),
  });
}

/**
 * AI 优化技能
 */
export async function refineSkill(
  skillId: string,
  feedback: string
): Promise<SkillGenerateResponse> {
  return apiRequest<SkillGenerateResponse>(`${BASE_URL}/${skillId}/refine`, {
    method: "POST",
    body: JSON.stringify({ feedback }),
  });
}

/**
 * 获取 Agent 的技能列表
 */
export async function getAgentSkills(
  agentId: string
): Promise<AgentSkillRead[]> {
  return apiRequest<AgentSkillRead[]>(`${BASE_URL}/agents/${agentId}`);
}

/**
 * 更新 Agent 的技能配置
 */
export async function updateAgentSkills(
  agentId: string,
  skills: AgentSkillConfig[]
): Promise<void> {
  await apiRequest<void>(`${BASE_URL}/agents/${agentId}`, {
    method: "PUT",
    body: JSON.stringify({ skills }),
  });
}

/**
 * 获取系统技能列表
 */
export async function getSystemSkills(): Promise<Skill[]> {
  return apiRequest<Skill[]>(`${BASE_URL}/system/list`);
}

/**
 * 初始化系统技能
 */
export async function initSystemSkills(): Promise<{ created: number }> {
  return apiRequest<{ created: number }>(`${BASE_URL}/system/init`, {
    method: "POST",
  });
}

/**
 * 重新加载技能缓存
 */
export async function reloadSkillCache(): Promise<void> {
  await apiRequest<void>(`${BASE_URL}/cache/reload`, {
    method: "POST",
  });
}

/**
 * 清除技能缓存
 */
export async function clearSkillCache(): Promise<void> {
  await apiRequest<void>(`${BASE_URL}/cache`, {
    method: "DELETE",
  });
}

// ========== 辅助函数 ==========

export const SKILL_TYPE_LABELS: Record<SkillType, string> = {
  system: "系统内置",
  user: "用户创建",
  ai: "AI 生成",
};

export const SKILL_CATEGORY_LABELS: Record<SkillCategory, string> = {
  prompt: "提示词增强",
  retrieval: "检索增强",
  tool: "工具扩展",
  workflow: "工作流",
};

export const AGENT_TYPE_OPTIONS = [
  { value: "product", label: "商品推荐" },
  { value: "faq", label: "FAQ" },
  { value: "knowledge", label: "知识库" },
  { value: "support", label: "客服" },
];

File: backend/app/routers/skills.py
"""技能管理 API 路由"""

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.dependencies import get_db_session
from app.core.logging import get_logger
from app.models.skill import SkillCategory, SkillType
from app.schemas.skill import (
    AgentSkillRead,
    AgentSkillsUpdate,
    SkillCreate,
    SkillGenerateRequest,
    SkillGenerateResponse,
    SkillListResponse,
    SkillRead,
    SkillRefineRequest,
    SkillUpdate,
)
from app.services.skill.generator import SkillGenerator
from app.services.skill.registry import skill_registry
from app.services.skill.service import SkillService
from app.services.skill.system_skills import SYSTEM_SKILLS

logger = get_logger("routers.skills")

router = APIRouter(
    prefix="/api/v1/admin/skills",
    tags=["skills"],
)


# ========== 技能 CRUD ==========


@router.get("", response_model=SkillListResponse)
async def list_skills(
    type: SkillType | None = None,
    category: SkillCategory | None = None,
    is_active: bool | None = None,
    page: int = 1,
    page_size: int = 20,
    db: AsyncSession = Depends(get_db_session),
):
    """获取技能列表"""
    service = SkillService(db)
    return await service.list_skills(
        skill_type=type,
        category=category,
        is_active=is_active,
        page=page,
        page_size=page_size,
    )


@router.post("", response_model=SkillRead, status_code=status.HTTP_201_CREATED)
async def create_skill(
    data: SkillCreate,
    db: AsyncSession = Depends(get_db_session),
):
    """创建技能"""
    service = SkillService(db)
    skill = await service.create_skill(data)
    await db.commit()
    await skill_registry.reload()
    return SkillRead.model_validate(skill)


@router.get("/{skill_id}", response_model=SkillRead)
async def get_skill(
    skill_id: str,
    db: AsyncSession = Depends(get_db_session),
):
    """获取技能详情"""
    service = SkillService(db)
    skill = await service.get_skill(skill_id)
    if not skill:
        raise HTTPException(status_code=404, detail="技能不存在")
    return SkillRead.model_validate(skill)


@router.put("/{skill_id}", response_model=SkillRead)
async def update_skill(
    skill_id: str,
    data: SkillUpdate,
    db: AsyncSession = Depends(get_db_session),
):
    """更新技能（系统技能不可更新）"""
    service = SkillService(db)
    try:
        skill = await service.update_skill(skill_id, data)
        if not skill:
            raise HTTPException(status_code=404, detail="技能不存在")
        await db.commit()
        await skill_registry.reload()
        return SkillRead.model_validate(skill)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.delete("/{skill_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_skill(
    skill_id: str,
    db: AsyncSession = Depends(get_db_session),
):
    """删除技能（系统技能不可删除）"""
    service = SkillService(db)
    try:
        success = await service.delete_skill(skill_id)
        if not success:
            raise HTTPException(status_code=404, detail="技能不存在")
        await db.commit()
        await skill_registry.reload()
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


# ========== AI 生成 ==========


@router.post("/generate", response_model=SkillGenerateResponse)
async def generate_skill(data: SkillGenerateRequest):
    """AI 生成技能（不需要数据库）"""
    generator = SkillGenerator()
    try:
        return await generator.generate(data)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/{skill_id}/refine", response_model=SkillGenerateResponse)
async def refine_skill(
    skill_id: str,
    data: SkillRefineRequest,
    db: AsyncSession = Depends(get_db_session),
):
    """AI 优化技能"""
    service = SkillService(db)
    skill = await service.get_skill(skill_id)
    if not skill:
        raise HTTPException(status_code=404, detail="技能不存在")

    # 转换为字典
    skill_data = {
        "name": skill.name,
        "description": skill.description,
        "category": skill.category.value,
        "content": skill.content,
        "trigger_keywords": skill.trigger_keywords,
        "trigger_intents": skill.trigger_intents,
        "always_apply": skill.always_apply,
        "applicable_agents": skill.applicable_agents,
    }

    generator = SkillGenerator()
    try:
        return await generator.refine(skill_data, data.feedback)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


# ========== Agent 技能配置 ==========


@router.get("/agents/{agent_id}", response_model=list[AgentSkillRead])
async def get_agent_skills(
    agent_id: str,
    db: AsyncSession = Depends(get_db_session),
):
    """获取 Agent 的技能列表"""
    service = SkillService(db)
    return await service.get_skills_for_agent(agent_id, include_disabled=True)


@router.put("/agents/{agent_id}", status_code=status.HTTP_204_NO_CONTENT)
async def update_agent_skills(
    agent_id: str,
    data: AgentSkillsUpdate,
    db: AsyncSession = Depends(get_db_session),
):
    """更新 Agent 的技能配置"""
    service = SkillService(db)
    await service.update_agent_skills(agent_id, data.skills)
    await db.commit()


# ========== 系统技能 ==========


@router.get("/system/list", response_model=list[SkillRead])
async def get_system_skills(
    db: AsyncSession = Depends(get_db_session),
):
    """获取系统内置技能"""
    service = SkillService(db)
    skills = await service.get_system_skills()
    return [SkillRead.model_validate(s) for s in skills]


@router.post("/system/init", status_code=status.HTTP_200_OK)
async def init_system_skills(
    db: AsyncSession = Depends(get_db_session),
):
    """初始化系统内置技能"""
    service = SkillService(db)
    count = await service.init_system_skills(SYSTEM_SKILLS)
    await db.commit()
    await skill_registry.reload()
    return {"message": f"初始化完成，新增 {count} 个技能", "created": count}


# ========== 缓存管理 ==========


@router.post("/cache/reload", status_code=status.HTTP_204_NO_CONTENT)
async def reload_skill_cache():
    """重新加载技能缓存"""
    await skill_registry.reload()


@router.delete("/cache", status_code=status.HTTP_204_NO_CONTENT)
async def clear_skill_cache():
    """清除技能缓存"""
    skill_registry.invalidate()
